lib/AST/ASTImporter.cpp

```diff
@@ -2534,7 +2571,7 @@ Decl *ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
               D->hasExternalFormalLinkage()) {
             if (IsStructuralMatch(D, FoundFunction)) {
               if (D->doesThisDeclarationHaveABody() &&
-                  FoundFunction->doesThisDeclarationHaveABody())
+                  FoundFunction->hasBody())
```

Comment:
Why is this changed?

Reply:
First I used `doesThisDeclarationHaveABody` in the case of function template specializations too.
But, there is a new test case (`ImportDefinitions`) which revealed an essential problem: `doesThisDeclarationHaveABody` is true only if that `FoundFunction` has a body, `hasBody` is true if there is one declaration with a body in the redecl chain of `FoundFunction`.
During parsing, `DeclContext::addDecl` will not add a new element to the StoredDeclsList rarther it simply overwrites the old element with the most recent one (see #386 ). And this most recent  decl will be found by lookup. However, the most recent decl can be a prototype without a body and in that case we will duplicate the `FunctionDecl`. Perhaps I could write a test for this case as well ...

Reply:
Is it possible that the `hasBody` will return true but the actual `FoundFunction` has no body (it exists in another Decl)? In this case a to-be imported FunctionDecl that has the body will be mapped to one that has no body (at that Decl). But I am not sure if this is problem.